CHANGELOG.md

```diff
@@ -1,10 +1,41 @@
 # HDMF Changelog
 
+## HDMF 4.0.0 (Future)
+
+### New Features
+- Added utility functions and classes useful for implementation of I/O backend to eases maintainability and reuse of functionality across I/O backends. @oruebel (#697)
+  - Added `HDF5IODataChunkIteratorQueue` class in `hdmf.backends.hdf5.h5_utils` as a new helper class for managing the write of `DataChunkIterators` to HDF5 @oruebel (#697)
+  - Added new `hdmf/io/utils.py` module with new I/O utility classes: 1) `WriteStatusTracker` as a simple data structure for tracking the write status of `Builders`, and 2) `NamespaceToBuilderHelper` to help with converting of a namespace to a `Builder` for I/O. @oruebel (#697)
+- Added `get_min_bounds` function to `hdmf.data_utils.DataChunk`. This functionality was originally part of `HDF5IO.__write_chunk__()` and has been moved here to enable reuse of the code across data backends. @oruebel (#697)
+- Added `ignore_string_to_byte` option for `TestCase.assertContainerEqual` to allow us to ignore conversion from string to bytes and just compare unicode values, as different backends may store strings differently. @oruebel (#697)
+
+### Enhancements
+- Updated `HDF5IO` to use the new `WriteStatusTracker`,`NamespaceToBuilderHelper`, and `HDF5IODataChunkIteratorQueue` helper classes. @oruebel (#697)
+  - Removed the following internal helper functions from `HDF5IO` as the functionality is now implemented by the helper classes: `HDF5IO.__convert_namespace`, `HDF5IO.__get_name`, `HDF5IO.__copy_spec`, `HDF5IO.__get_new_specs`, `HDF5IO.__builderhash`, `HDF5IO.__exhaust_dcis`, `HDFIO.__write_chunk__`
+- Added `hdmf.utils.__get_docval_macros` function and updated `hmdf.testing.testcase.TestCase` to use `__get_docval_macros`  to determine array types to remove dependency on h5py and allow support for other array types (e.g., from other I/O backends) in testing. @oruebel (#697)
+- Updated `hdmf.common.sparse.CSRMatrix` to avoid direct dependency on h5py as a particular storage backend. @oruebel (#697)
+
+### Bug fixes
+- Updated `HDF5IO` to always set the `location` attribute of `GroupBuilders`, `DatasetBuilders`, and `LinkBuilders` on read.   @oruebel (#697)
+- Updated `HDF5IO.__get_path` to use `Builder.location` if available.  @oruebel (#697)
+- Updated `HDF5IO.get_types` to correctly determine the data type for `bytes` data. @oruebel (#697)
+
+### Breaking changes
+- Moved `UnsupportedOperation` error class from `hdmf.backends.io` to the new `hdmf.backends.errors` module.  @oruebel (#697)
+- Moved `tests.unit.test_io_hdf5_h5tools.FooFile` to `tests.unit.utils.FooFile` to ease reuse and to consolidate the `Foo` test data classes in one module.  @oruebel (#697)
```

Comment:
Breaking changes in tests do not break compatibility in the API so I would consider this a non-breaking change